core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java

```diff
@@ -0,0 +1,438 @@
+   @Inject
+   public void inject(AsyncInterceptorChain interceptorChain,
+                      AdvancedCache<K, V> cache,
+                      CommandsFactory commandsFactory,
+                      DistributionManager distributionManager,
+                      EntryMergePolicy<K, V> entryMergePolicy,
+                      @ComponentName(STATE_TRANSFER_EXECUTOR) ExecutorService stateTransferExecutor,
+                      InvocationContextFactory invocationContextFactory,
+                      KeyPartitioner keyPartitioner,
+                      RpcManager rpcManager,
+                      StateConsumer stateConsumer,
+                      StateReceiver<K, V> stateReceiver) {
+      this.interceptorChain = interceptorChain;
+      this.cache = cache;
+      this.commandsFactory = commandsFactory;
+      this.distributionManager = distributionManager;
+      this.entryMergePolicy = entryMergePolicy;
+      this.stateTransferExecutor = stateTransferExecutor;
+      this.invocationContextFactory = invocationContextFactory;
+      this.keyPartitioner = keyPartitioner;
+      this.rpcManager = rpcManager;
+      this.stateConsumer = stateConsumer;
+      this.stateReceiver = stateReceiver;
+   }
+
+   @Start
+   public void start() {
+      this.cache = cache.withFlags(Flag.IGNORE_RETURN_VALUES, Flag.PUT_FOR_STATE_TRANSFER);
+      this.cache.addListener(this);
+      this.cacheName = cache.getName();
+      this.localAddress = rpcManager.getAddress();
+
+      initRpcOptions();
+      cache.getCacheConfiguration().clustering()
+            .attributes().attribute(ClusteringConfiguration.REMOTE_TIMEOUT)
+            .addListener(((a, o) -> initRpcOptions()));
+   }
+
+   private void initRpcOptions() {
+      this.rpcOptions = rpcManager.getRpcOptionsBuilder(ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS).build();
+   }
+
+   @TopologyChanged
+   @SuppressWarnings("unused")
+   public void onTopologyChanged(TopologyChangedEvent<K, V> event) {
+      // We have to manually create a LocalizedTopology here, as CacheNotifier::notifyTopologyChanged is called before
+      // DistributionManager::setTopology and we always need to utilise the latest hash
+      if (trace) log.tracef("Installed new topology %s: %s", event.getNewTopologyId(), event.getWriteConsistentHashAtEnd());
+      this.installedTopology = distributionManager.getCacheTopology();
+   }
+
+   @DataRehashed
+   @SuppressWarnings("unused")
+   public void onDataRehashed(DataRehashedEvent<K, V> event) {
+      if (event.isPre() && isStateTransferInProgress()) {
+         // We know that a rehash is about to occur, so postpone all requests which are affected
+         synchronized (versionRequestMap) {
+            versionRequestMap.values().forEach(VersionRequest::cancelRequestIfOutdated);
+         }
+      } else if (!event.isPre() && event.getPhase() == CacheTopology.Phase.READ_ALL_WRITE_ALL) {
+         this.installedTopology = createLocalizedTopology(event.getConsistentHashAtEnd());
```

Comment:
I've changed how this works now, so that the `StateConsumerImp` interacts directly with the `DefaultConflictManager`.